pandas/core/indexes/base.py

```diff
@@ -2862,7 +2862,12 @@ def _union(self, other, sort):
         lvals = self._values
         rvals = other._values
 
-        if sort is None and self.is_monotonic and other.is_monotonic:
+        if (
+            sort is None
+            and self.is_monotonic
+            and other.is_monotonic
+            and not (self.has_duplicates and other.has_duplicates)
+        ):
```

Comment:
can you change this routine to just return result as soon as its possible. this will make it easiest to grok. e.g. I think in the try on L2872 you can just return (if it succeeds). and I think L2880 maybe as well (i guess you might nee dto sort here not sure).

Reply:
IOW i think its ok to duplicate the _maybe_try_sort if we can early exit this function w/o this giant if/else with many branches

Reply:
another is to make this really if/elif/elif/else. that would be fine as well. This nested if/else is just hard to grok.

Reply:
Reformatted it. Hope it is better readable now. Mixed both approaches